adr/ADR-13.md

```diff
@@ -0,0 +1,290 @@
+        "ack_policy": "explicit",
+        "ack_wait": 30000000000,
+        "max_deliver": -1,
+        "filter_subject": "bar",
+        "replay_policy": "instant",
+        "max_waiting": 512,          <-- Maximum Inflight Pull/Fetch Requests [1]
+        "max_ack_pending": 20000
+    },
+    "delivered": {
+        "consumer_seq": 11561,
+        "stream_seq": 11560
+    },
+    "ack_floor": {
+        "consumer_seq": 11561,
+        "stream_seq": 11560
+    },
+    "num_ack_pending": 0,
+    "num_redelivered": 0,
+    "num_waiting": 1,                <-- Inflight Pull/Fetch Requests [2]
+    "num_pending": 0,
+    "cluster": {
+        "leader": "NCUOUH6YICRESE73CKTXESMGA4ZN4KTELXPUIE6JCRTL6IF4UWE3B2Z4"
+    }
+}
+```
+
+When making a pull request it is also possible to request more than one message:
+
+```shell
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.x7tkDPDLCOEknrfB4RH1V7.OgY4M7 11
+{"batch":5}
+```
+
+Whenever a pull request times out, the count of `num_waiting` will increase for a consumer
+but this will eventually reset once it reaches the max waiting inflight that was configured
+for the pull consumer.
+
+### No Wait Pull Requests
+
+In order to get a response from the server rigth away, a client can
+make a pull request with the `no_wait` option enabled. For example:
+
+```shell
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.x7tkDPDLCOEknrfB4RH1V7.OgY4M7 26
+{"batch":1,"no_wait":true}
+```
+
+The result of a no wait pull request is a guaranteed instant response
+from the server that will be either the next message or an error,
+where an error could be a server error such as `503` in case JS
+service is not available. Most commonly, the result of a no wait pull
+request will be a `404` no messages error:
+
+```shell
+HMSG _INBOX.x7tkDPDLCOEknrfB4RH1V7.OgY4M7 2  28 28
+NATS/1.0 404 No Messages
+```
+
+## Design
+
+The implementation for pull subscribe uses a combination of both no wait and
+lingering pull requests described previously.
+
+In the Go client, a simple example of the API looks as follows:
+
+```go
+sub, err := js.PullSubscribe("stream-name", "durable")
+if err != nil {
+	log.Fatal(err)
+}
+
+for {
+	msgs, err := sub.Fetch(1)
+	if err != nil {
+		log.Println("Error:", err)
+		continue
+	}
+	for _, msg := range msgs {
+		msg.Ack()
+	}
+}
+```
+
+When implementing `PullSubscribe` there are two main cases to
+consider: `Fetch(n)` and `Fetch(1)`.
+
+#### Pull(n)
+
+`Pull(n)` batch requests are implemented somewhat similarly to old style
+requests.  When making a pull request, first a no wait request is
+done to try to get the messages that may already be available as
+needed.  If there are no messages (a 404 status message error by the
+JetStream server), then a long pull request is made:
+
+```shell
+SUB _INBOX.NQ2uAIXd4GoozKOTfECtIg  1
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 26
+{"batch":5,"no_wait":true}
+
+# Result of no wait request if there are no messages
```

Comment:
it would then wait for all the messages in the batch to be delivered or the request will timeout waiting for the complete batch of messages.

Reply:
Ok, so `no_wait` first says there are no messages, then we go to "long" wait and again batch is 5, but only 4 are avail, so library will wait up to the timeout and then will return no error since there is at least 1 message to return to the user.
(what if there is no timeout specified, do we wait for ever? in here we have 4 messages out of 5 so no message will ever be returned. So in other words, is a timeout mandatory for Fetch() calls?)

Reply:
By default in nats.go the timeout is 5 seconds, but users can pass a context without timeout and drive the cancellation on their own, etc... but thinking that maybe we should make the timeout mandatory.

Reply:
Some other reasons that `no_wait` is used as well is to further help for the `Pull(1)` use case (which likely is the most common), 'no wait' requests do not become long pull requests, run into 408 Request Timeout errors since not accounted as inflight pull request since meant to be instant.
Also just found a mistake in the unsub, it should have been `+1`...

Reply:
fixed now